sdk/python/kfp/compiler/compiler_utils.py

```diff
@@ -427,14 +428,42 @@ def get_dependencies(
                 task2=task,
             )
 
-            # a task cannot depend on a task created in a for loop group since individual PipelineTask variables are reassigned after each loop iteration
-            dependent_group = group_name_to_group.get(upstream_groups[0], None)
-            if isinstance(dependent_group,
-                          (tasks_group.ParallelFor, tasks_group.Condition,
-                           tasks_group.ExitHandler)):
+            # uncommon upstream ancestor check
+            uncommon_upstream_groups = deepcopy(upstream_groups)
+            uncommon_upstream_groups.remove(
+                upstream_task.name
+            )  # because a task's `upstream_groups` contains the task's name
+            if uncommon_upstream_groups:
+                dependent_group = group_name_to_group.get(
+                    upstream_groups[0], None)
+                if isinstance(dependent_group, tasks_group.ExitHandler):
+                    task_group_type = 'an ' + tasks_group.ExitHandler.__name__
+
+                elif isinstance(dependent_group, tasks_group.Condition):
+                    task_group_type = 'a ' + tasks_group.Condition.__name__
+
+                else:
+                    task_group_type = 'a ' + tasks_group.ParallelFor.__name__
+
                 raise RuntimeError(
-                    f'Task {task.name} cannot dependent on any task inside'
-                    f' the group: {upstream_groups[0]}.')
+                    f'Tasks cannot depend on upstream tasks inside {task_group_type} that is not a common ancestor of both tasks. Task {task.name} depends on upstream task {upstream_task.name}.'
+                )
+
+            # ParralelFor Nested Check
+            # if there is a parrallelFor group type in the upstream parents tasks and there also exists a parallelFor in the uncommon_ancestors of downstream: this means a nested for loop exists in the DAG
+            upstream_parent_tasks = task_name_to_parent_groups[
+                upstream_task.name]
+            for group in downstream_groups:
+                if isinstance(
+                        group_name_to_group.get(group, None),
+                        tasks_group.ParallelFor):
+                    for parent_task in upstream_parent_tasks:
+                        if isinstance(
+                                group_name_to_group.get(parent_task, None),
+                                tasks_group.ParallelFor):
+                            raise RuntimeError(
+                                f'Downstream task cannot depend on an upstream task while in a nested {tasks_group.ParallelFor.__name__} group. Task {task.name} depends on upstream task {upstream_task.name}, while {group} is nested in {parent_task}'
```

Comment:
I think this case this handles is actually slightly narrower than the error message suggests. The issue is not that the downstream task cannot be in a nested `ParallelFor` since the following is allowed:

```python
@dsl.pipeline
def p():
    t1 = c1()
    with dsl.ParallelFor([1,2,3]):
        with dsl.ParallelFor([1,2,3]):
            t1 = c1(x=t1.output)
```

What do you think about athe following revised first sentence:

`'Downstream tasks in a nested ParallelFor cannot depend on an upstream task in a shallower ParallelFor.`

Still leaves a bit more precision to be desired, but I think it captures the idea.

Reply:
nit: also add period to the end of the message for consistency with the above error message